### PR TITLE
c8d/prune: Keep the last tagged image instead of creating dangling image

### LINKED ISSUE #48063: containerd integration: Prune deletes images used by containers created from a digested reference
### Description

`docker image prune -a` will untag used image when the only container using the image was created from an image specified by ID.

Additionally, when the container is started from a digested reference (like `busybox@sha256:abcdef...`) prune will not only untag the image, but also delete the image content.

While the first behavior might make sense - it's inconsistent with the graphdrivers implementation.

### Reproduce

```bash
$ docker pull alpine

$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED      SIZE
alpine       latest    b89d9c93e9ed   5 days ago   13.6MB

$ docker run -d alpine@sha256:b89d9c93e9ed3597455c90a0b88a8bbb5cb7188438f70953fede212a0c4394e0 sleep infinity
...

$ docker image prune -a
WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: alpine:latest
deleted: sha256:b89d9c93e9ed3597455c90a0b88a8bbb5cb7188438f70953fede212a0c4394e0
deleted: sha256:647a509e17306d117943c7f91de542d7fb048133f59c72dff7893cffd1836e11
deleted: sha256:092561eea88f9f28654bbade209576f5f93efeb8c7ba66a07ac2033c0ddc8ae7
deleted: sha256:a258b2a6b59a7aa244d8ceab095c7f8df726f27075a69fca7ad8490f3f63148a
```

```bash
$ docker pull alpine

$ docker images
REPOSITORY   TAG       IMAGE ID       CREATED      SIZE
alpine       latest    b89d9c93e9ed   5 days ago   13.6MB

$ docker run -d sha256:b89d9c93e9ed3597455c90a0b88a8bbb5cb7188438f70953fede212a0c4394e0 sleep infinity
...

$ docker image prune -a
WARNING! This will remove all images without at least one container associated to them.
Are you sure you want to continue? [y/N] y
Deleted Images:
untagged: alpine:latest
```

### Expected behavior

Prune definitely shouldn't delete content of images used by containers.
Prune **probably**  shouldn't (if we want to be consistent with graphdrivers) untag images if used by containers that were started off the image ID.

### docker version

```bash
Client:
 Version:           27.0.1
 API version:       1.46
 Go version:        go1.21.11
 Git commit:        7fafd33
 Built:             Mon Jun 24 14:56:52 2024
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.33.0 (1)
 Engine:
  Version:          27.0.1
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.21.11
  Git commit:       ff1e2c0
  Built:            Mon Jun 24 14:58:04 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.7.18
  GitCommit:        ae71819c4f5e67bb4d5ae76a6b735f29cc25774e
 runc:
  Version:          1.7.18
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
containerd integration enabled
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by vvoland**
- fixes: https://github.com/moby/moby/issues/48063

Don't turn images into dangling when they are used by containers created with an image specified by an ID only (e.g. `docker run 82d1e9d`).

Keep the last image reference with the same target when all other references would be pruned.

If the container was created with a digested and tagged reference (e.g. `docker run alpine:latest@sha256:82d1e9d7ed48a7523bdebc18cf6290bdb97b82302a8a9c27d4fe885949ea94d1`), the `alpine:latest` image won't get untagged.

This change makes the behavior consistent with the graphdriver implementation.

**- What I did**

**- How I did it**

**- How to verify it**
TestPruneDontDeleteUsedImage

**- Description for the changelog**

```markdown changelog
containerd integration: Fix `docker image prune -a` untagging images used by containers started from images referenced by a digested reference.
```

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by vvoland** [comment]
Hmm, getting weird failures to start secondary daemon (overlay2 not supported?) in integration tests:
```

2024-06-27T16:05:08.9359404Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.344847699Z" level=info msg="Starting up"
2024-06-27T16:05:08.9362497Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.345499908Z" level=debug msg="Listener created for HTTP on unix (/tmp/docker-integration/df78e0f4df204.sock)"
2024-06-27T16:05:08.9364782Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.354504908Z" level=debug msg="Golang's threads limit set to 115020"
2024-06-27T16:05:08.9366455Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.354997418Z" level=debug msg="metrics API listening on /tmp/dxr/df78e0f4df204/metrics.sock"
2024-06-27T16:05:08.9367646Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.357471248Z" level=debug msg="Using default logging driver json-file"
2024-06-27T16:05:08.9368878Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.357530727Z" level=debug msg="processing event stream" module=libcontainerd namespace=df78e0f4df204p
2024-06-27T16:05:08.9371322Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.357662128Z" level=debug msg="No quota support for local volumes in /go/src/github.com/docker/docker/bundles/test-integration/TestPruneDontDeleteUsedImage/single_tag/repo_and_digest,_no_tag/df78e0f4df204/root/volumes: Filesystem does not support, or has not enabled quotas"
2024-06-27T16:05:08.9373348Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.363082482Z" level=info msg="[graphdriver] trying configured driver: overlay2"
2024-06-27T16:05:08.9374572Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.380383275Z" level=error msg="failed to mount overlay: invalid argument" storage-driver=overlay2
2024-06-27T16:05:08.9375843Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.380407951Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
2024-06-27T16:05:08.9377146Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.380416497Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
2024-06-27T16:05:08.9378309Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.381139810Z" level=debug msg="Cleaning up old mountid : start."
2024-06-27T16:05:08.9379322Z     daemon.go:317: [df78e0f4df204] time="2024-06-27T16:04:04.381418636Z" level=debug msg="Cleaning up old mountid : done."
2024-06-27T16:05:08.9380267Z     daemon.go:317: [df78e0f4df204] failed to start daemon: error initializing graphdriver: driver not supported: overlay2
2024-06-27T16:05:08.9384340Z     prune_test.go:152: [df78e0f4df204] failed to start daemon with arguments [--data-root /go/src/github.com/docker/docker/bundles/test-integration/TestPruneDontDeleteUsedImage/single_tag/repo_and_digest,_no_tag/df78e0f4df204/root --exec-root /tmp/dxr/df78e0f4df204 --pidfile /go/src/github.com/docker/docker/bundles/test-integration/TestPruneDontDeleteUsedImage/single_tag/repo_and_digest,_no_tag/df78e0f4df204/docker.pid --userland-proxy=true --containerd-namespace df78e0f4df204 --containerd-plugins-namespace df78e0f4df204p --containerd /var/run/docker/containerd/containerd.sock --host unix:///tmp/docker-integration/df78e0f4df204.sock --debug --storage-driver overlay2] : [df78e0f4df204] daemon exited during startup: exit status 1
```

They work fine locally 🤔

**Comment by laurazard** [comment]
I was testing this on graphdrivers vs containerd, and noticed on graphdrivers, we seem to always remove the last tag alphanumerically, but with the containerd implementation it seems a bit random.

I wrote a repro here:

```bash
#!/usr/bin/env bash
set -ex
while true; do
    docker image prune -af
    docker pull busybox:latest
    repoDigest=$(docker inspect busybox:latest | jq .[0].RepoDigests[0] | tr -d '"')
    docker tag busybox:latest busybox:a
    docker tag busybox:latest busybox:b
    docker tag busybox:latest busybox:c
    docker rmi busybox:latest
    ctrId=$(docker run -d $repoDigest sleep 100)
    yes | docker image prune -a
    tags=$(docker images --filter=reference='busy*' --format json | jq .Tag | tr -d '"')
    tagsLength=${#tags[@]}
    if [[ $tagsLength -ne 1 ]]; then
        echo "this shouldn't happen! tags length != 1"
        exit 1
    fi
    if [[ $tags != c ]]; then
        exit 1
    fi
    docker kill $ctrId
    docker rm $ctrId
done
```

**Comment by laurazard** [comment]
I don't know whether **must** keep this behavior, but it sounds safer 😅.

File: integration/image/prune_test.go [hunk 1]
```diff
@@ -47,3 +52,172 @@ func TestPruneDontDeleteUsedDangling(t *testing.T) {
...
+	// run container
+	cid := container.Run(ctx, t, apiClient, container.WithImage(id))
+	defer container.Remove(ctx, t, apiClient, cid, containertypes.RemoveOptions{Force: true})
+
+	pruned, err := apiClient.ImagesPrune(ctx, filters.NewArgs(filters.Arg("dangling", "false")))
+	assert.NilError(t, err)
+
+	assert.Check(t, is.Len(pruned.ImagesDeleted, 1))
+	assert.Check(t, is.Equal(pruned.ImagesDeleted[0].Deleted, ""))
+	assert.Check(t, is.Equal(pruned.ImagesDeleted[0].Untagged, "busybox:a"))
```
[the remark this answers is not in the document]
Comment (vvoland) [reply integration/image/prune_test.go]: Changed to 10 images tagged in non-lexographic order :)

**Review by laurazard — APPROVED** [review]
LGTM 🎉

File: daemon/containerd/image_prune.go [hunk 2]
```diff
@@ -107,8 +108,16 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 
 	usedDigests := filterImagesUsedByContainers(ctx, i.containers.List(), imagesToPrune)
 
+	// Sort images by name to make the behavior deterministic and consistent with graphdrivers.
+	sorted := make([]string, 0, len(imagesToPrune))
+	for name := range imagesToPrune {
+		sorted = append(sorted, name)
+	}
```
Comment (thaJeztah) [inline daemon/containerd/image_prune.go:115]: We can probably start looking at the `maps` package, and for these kind of things, use [`maps.Keys()`](https://pkg.go.dev/golang.org/x/exp/maps#Keys)

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Pull request merged**